### DISCUSSION TITLE
[QUESTION] Thinking about migrating from Next.js, need some insight

### DISCUSSION THREAD

**Discussion by alessandrojcm** [vikejs/vike#1460] [Discussion] [answered]
Hi there! So, for the past year, I've been leading a migration from an old legacy MVC app to a new React application. For this, I've chosen Next.js as I have been working with the framework for the past 3 years or so and have never had an issue; yes for sure some quirks that I have had to work around but nothing spectacular.

This application I am working on is a very niche financial product that gets deployed on-premise to the client's intranet using either k8s or Ansible; hence the app is running inside a container with the Node.js runtime. The migration is going to be a phased one, so we will implement a feature in React, then add a link to the old app that redirects to a route in the new React app until the migration is complete and we can deprecate the old app; this means that for the foreseeable we will have to support both applications.

So far so good, but then I stumbled on this: to support the legacy system, we will need to have a Redis cluster in place that needs to be reachable by the two apps to handle sessions, I am handling authentication inside middleware with Auth.js because that is the most natural place to do it. Now, due to an [arbitrary limitation](https://github.com/vercel/next.js/discussions/34179) imposed by Vercel when you're running on the Edge (which we are **not** doing), I cannot span a TCP connection from middleware. The solution? Use either Vercel KV or Upstash, something we _cannot_ do because, as I mentioned, we are deploying on-prem inside an intranet. 

Looking around I have found a [random Docker image](https://github.com/hiett/serverless-redis-http) to deploy a service, written in Elixir, that mocks the Upstash API 1:1 so the recommended solution by Vercel can be run completely within our deployment model. Now, I think this is just too much, I have to go to the engineering manager and the architect and explain to them that we need to provision another piece of infra that I just happen to have found randomly on the internet which is also written in a language none of us have experience with, effectively introducing another point of failure, just to work around a random limitation imposed by the framework. This and [random regression bugs](https://github.com/vercel/next.js/issues/60890#issuecomment-1900593533) that get introduced with minor or patch versions is just a bit too much for me, especially this early in the development cycle.

Fortunately, most of the code that is in place is pure React code that does not make much use of anything Next.js specific; beyond the routing structure. Hence, I am considering the effort of just ditching Next.js altogether. Vike seems like a good solution, and I have no problem spending a bit more time writing boilerplate if that means flexibility in the long run. So, here is a list of the features that I am using and I would like to get more insight if and how could I replicate these with Vike:

* **Route Handlers**: I understand that Vike [does not support API Routes](https://vike.dev/api-routes) and the recommended approach is to use Telefunc. That is fine by me but I do have a couple of route handlers so I´d like to know if they are doable using Telefunc:
    * One of them is an endpoint that generates CSV of PDF files
    * Another one is to query the database since Sequelize does not run on the edge (I guess with Vike this would not be necessary as I can just query the db directly)
    * An endpoint that uses the Elasticsearch client to query an Elasticsearch cluster directly and returns search results
* **Nested Layouts**: I've read in the docs that this is supported out of the box but how different is it to Next's structure?
* **Authentication**: I am using Auth.js, so that should be framework agnostic. But does Vike support some sort of middleware so I can handle auth there?
* **I18n**: I am using `next-intl` and everything is mostly in place. This is also can be framework agnostic but if there is something with a similar API in Vike (ICU messages, global formatting of numbers/dates) then that would be great
* **RSC**: this is the main pain point, I am fully sold with RSC, streaming, and Suspense, and I make good use of it. I know there is server streaming supported within Vike but is it possible to achieve something similar to Next's RSC/Suspense implementation?
* **SSR**: I am a bit lost here if I want to be able to do SSR and pass props down to components, do I need to couple it with my server? Is that correct?

Thanks in advance!

**Comment by nitedani** [comment, marked as answer]
Authentication, CSV and PDF files: you can add a regular api route/middleware in your custom server, in the followng file for example:
https://github.com/vikejs/vike/blob/main/examples/telefunc/server/index.js
Example using `express` and `@aws-sdk/client-s3`
```ts
app.get("/pdf", (req, res) => {
  // const result = ... get from s3
  const stream = Readable.fromWeb(
    result.Body.transformToWebStream() as ReadableStream
  );
  const type = result.ContentType;
  const length = result.ContentLength;
  if (type) {
    res.type(type);
  }
  if (length) {
    res.header("Content-Length", String(length));
  }

  res.status(200);
  stream.pipe(res);
});
```

query the database: telefunc supports this, you can write the query with telefunc
Elasticsearch: same, you can write it with telefunc

something similar to Next's RSC/Suspense implementation: [RSC is not yet implemented](https://vike.dev/react#react-server-components) in `vike-react`, but there is something cool you can use instead: [vike-react-query](https://github.com/vikejs/vike-react/tree/main/packages/vike-react-query)
Example: https://github.com/vikejs/vike-react/blob/main/examples/react-query/pages/index/%40id/%2BPage.tsx
This would stream the initial data on component level, making the independent components interactive on client as soon as the data is available

I18n: you would need to integrate the chosen library yourself (for example adding a context provider, saving the chosen language in a cookie/parsing it from headers/from the route string)

**Reply by alessandrojcm** [reply]
Hi there, thanks for the examples that's very useful! About `vite-react-query`, so this basically mimics the Loading/Error/Suspended component pattern from Next.js am I correct? The loading component gets render first while the main component is streaming when it's ready? Can I pass props from the server side with this? Also is there something similar to the page level error boundaries? (ie Error.tsx)

**Reply by nitedani** [reply]
The loading component is rendered first, if the data for the given queryKey is not yet available in the queryClient cache. The queries are also deduplicated by the queryKey. The queryClient cache is separate per request on the server, and global on the client. If there is an error during loading/rendering the component, the Error is rendered.

Error boundary: you can define page-level error boundaries, for example using `react-error-boundary`:

```tsx
// error.tsx
import { ErrorBoundary } from 'react-error-boundary'

export const FallbackErrorBoundary = ({ children }) => (
  <ErrorBoundary onReset={() => {}} fallback={<div>error</div>}>
    {children}
  </ErrorBoundary>
)
```

When using vike-react-query, you can set it in the page config
```ts
//+config.h.ts
import { FallbackErrorBoundary } from './error'

export default {
  ...
  FallbackErrorBoundary
} satisfies Config
```

Alternatively, you can use it directly in the page component:
```tsx
import { FallbackErrorBoundary } from './error'

function Page() {
  return <FallbackErrorBoundary>...</FallbackErrorBoundary>
}
```


Can I pass props from the server side with this?: what do you exactly mean by this? You can combine telefunc and vike-react-query, then, the query function will always run on the server:

```tsx
 // component.tsx
const result = useSuspenseQuery({
  queryKey: ['movie', id],
  queryFn: () => getStarWarsMovie(id),
})
```

```ts
 //component.telefunc.ts
async function getStarWarsMovie(id: string): Promise<MovieDetails> {
  
  fs.readFile(...)

  const response = await fetch(`https://star-wars.brillout.com/api/films/${id}.json`)
  return response.json()
}
```

**Reply by alessandrojcm** [reply]
Thans, that's very helpful. Will take a look at everything you suggested

**Reply by brillout** [reply]
> * **RSC**: [...] I am fully sold with RSC, streaming, and Suspense, and I make good use of it. I know there is server streaming supported within Vike but is it possible to achieve something similar to Next's RSC/Suspense implementation?

We also like RSC a lot. That said, the one aspect we're a bit relunctant about RSC are server-only comopnents. While the idea is promising, I think it's a little too early for using it in production. That said, [Progressive Rendering](https://vike.dev/streaming#progressive-rendering) and RPC (with Telefunc instead of Server Actions) are already implemented. See https://vike.dev/react#react-server-components for more information.

> * **Nested Layouts**: I've read in the docs that this is supported out of the box but how different is it to Next's structure?

See https://vike.dev/Layout#nested-layouts. It's a little bit more boilerplate but, in exchange, you get a lot more flexibility.

> * **Authentication**: I am using Auth.js, so that should be framework agnostic. But does Vike support some sort of middleware so I can handle auth there?

Vike is decoupled from the server (it's a core Vike belief). So you integrate authentication direclty on your server, see https://vike.dev/auth.

> * **I18n**: I am using `next-intl` and everything is mostly in place. This is also can be framework agnostic but if there is something with a similar API in Vike (ICU messages, global formatting of numbers/dates) then that would be great

See https://vike.dev/i18n. As usual: a little bit more boilerplate but, in exchange, you get a lot more flexibility.

> * **SSR**: I am a bit lost here if I want to be able to do SSR and pass props down to components, do I need to couple it with my server? Is that correct?

I'm not sure what you mean.

---

I don't see any blockers for your use case.

You may have to write a little bit more boilerplate code to integrate tools, but it's often a good thing in the long run.

> [random regression bugs](https://github.com/vercel/next.js/issues/60890#issuecomment-1900593533)

FYI we [systematically and quickly fix bugs](https://github.com/vikejs/vike/issues?q=is%3Aissue+is%3Aclosed+label%3A%22bug+%3Aboom%3A%22).

Also FYI, you can [sponsor](https://github.com/vikejs/vike/issues/1350) and get tight-knit support from us. This has been working great with our sponsors (they build quite some advanced apps using Vike).

**Reply by alessandrojcm** [reply]
Thanks for your input @brillout, yeah I might have jumped the gun on RSC a bit too soon but luckily is early enough in the project to steer course. The `vike-react-query` package looks very interesting anyway and well encapsulates the Loading/Error/Component paradigm albeit in a more granular level

>I'm not sure what you mean.
>
Sorry this was poorly worded, anyway my point is can be accomplished using `vite-react-query` and telefunc.
> Also FYI, you can https://github.com/vikejs/vike/issues/1350 and get tight-knit support from us. This has been working great with our sponsors (they build quite some advanced apps using Vike).
> 
Good point, we're discussing options next week; I'll make sure to point this out.
